pkg/vpc/cleanup.go

```diff
@@ -30,49 +33,50 @@ func findAvailableNetworkInterfaces(provider api.ClusterProvider, spec *api.Clus
 			},
 		},
 	}
-	output, err := provider.EC2().DescribeNetworkInterfaces(input)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
-	}
+
 	securityGroupRE, err := regexp.Compile(fmtSecurityGroupNameRegexForCluster(spec.Metadata.Name))
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
+		return errors.Wrap(err, "failed to create security group regex")
 	}
-	networkInterfaces := []string{}
-	for _, eni := range output.NetworkInterfaces {
-		id := *eni.NetworkInterfaceId
-		for _, sg := range eni.Groups {
-			if securityGroupRE.MatchString(*sg.GroupName) {
-				logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
-				networkInterfaces = append(networkInterfaces, id)
-				break
-			} else {
+
+	var lastErr error
+
+	err = ec2API.DescribeNetworkInterfacesPagesWithContext(ctx, input, func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool {
+		for _, eni := range output.NetworkInterfaces {
+			id := *eni.NetworkInterfaceId
+			for _, sg := range eni.Groups {
+				if securityGroupRE.MatchString(*sg.GroupName) {
+					logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
+					if err := fn(id); err != nil {
+						lastErr = err
+						return false
+					}
+					break
```

Comment:
Yes, we need to find all of them!

Reply:
Good question. We stop because we only need to check if the network interface has at least one security group matching the regex. After invoking the callback function, the ENI would no longer exist and if another security group for the same ENI matches the regex, the callback function fail to delete the already-deleted ENI.